File: solutions/go/bob/2/bob.go
package bob

import (
    "strings"
    "unicode"
)

type Response = string

const (
    QuestionResponse 		Response = "Sure."
    YellingResponse			Response = "Whoa, chill out!"
    YellingQuestionResponse	Response = "Calm down, I know what I'm doing!"
    SilenceResponse			Response = "Fine. Be that way!"
    OtherwiseResponse		Response = "Whatever."
)

func Hey(remark string) Response {
    trimmed := strings.TrimSpace(remark)
    isYelling := strings.ContainsFunc(trimmed, hasLetter) && !strings.ContainsFunc(trimmed, hasLower)
    isAsking := strings.HasSuffix(trimmed, "?")
    switch {
    case len(trimmed) == 0:
        return SilenceResponse
    case isYelling && isAsking:
        return YellingQuestionResponse
    case isYelling:
        return YellingResponse
    case isAsking:
        return QuestionResponse
    default:
        return OtherwiseResponse
    }
}

func hasLetter(r rune) bool {
    return unicode.IsLetter(r)
}

func hasLower(r rune) bool {
    return unicode.IsLower(r)
}

File: solutions/typescript/bob/1/bob.ts
const RESPONSES = {
  Question: "Sure.",
  Yelling: "Whoa, chill out!",
  YellingQuestion: "Calm down, I know what I'm doing!",
  Silence: "Fine. Be that way!",
  Otherwise: "Whatever.",
} as const

type Response = typeof RESPONSES[keyof typeof RESPONSES]

export function hey(message: string): Response {
  const trimmed = message.trim()
  if (trimmed.length === 0) {
    return RESPONSES.Silence
  }
  const isAsking = trimmed.endsWith("?")
  const upper = trimmed.toUpperCase()
  const isYelling = trimmed.toLowerCase() !== upper && trimmed === upper
  switch (true) {
    case isYelling && isAsking:
      return RESPONSES.YellingQuestion
    case isYelling:
      return RESPONSES.Yelling
    case isAsking:
      return RESPONSES.Question
    default:
      return RESPONSES.Otherwise
  }
}